molgenis-data-semanticsearch/src/main/java/org/molgenis/data/semanticsearch/service/impl/SemanticSearchServiceHelper.java

```diff
@@ -0,0 +1,144 @@
+				"what", "what's", "when", "when's", "where", "where's", "which", "while", "who", "who's", "whom",
+				"why", "why's", "with", "won't", "would", "wouldn't", "you", "you'd", "you'll", "you're", "you've",
+				"your", "yours", "yourself", "yourselves", "many", ")", "("));
+	}
+
+	@Autowired
+	public SemanticSearchServiceHelper(OntologyTagService ontologyTagService, DataService dataService,
+			OntologyService ontologyService)
+	{
+		if (null == ontologyTagService || null == dataService || null == ontologyService) throw new MolgenisDataException(
+				"Service is not found, please contact your application administrator");
+
+		this.dataService = dataService;
+		this.ontologyTagService = ontologyTagService;
+		this.ontologyService = ontologyService;
+	}
+
+	public QueryRule createDisMaxQueryRule(EntityMetaData targetEntityMetaData, AttributeMetaData targetAttribute)
+	{
+		LinkedHashMultimap<Relation, OntologyTerm> tagsForAttribute = ontologyTagService.getTagsForAttribute(
+				targetEntityMetaData, targetAttribute);
+
+		List<QueryRule> rules = new ArrayList<QueryRule>();
+
+		// add query rule for searching the label of target attribute in the attribute table.
+		if (StringUtils.isNotEmpty(targetAttribute.getDescription()))
+		{
+			rules.add(new QueryRule(AttributeMetaDataMetaData.LABEL, Operator.FUZZY_MATCH, targetAttribute
+					.getDescription()));
+		}
+
+		for (OntologyTerm ontologyTerm : tagsForAttribute.values())
+		{
+			QueryRule disMaxQuery = new QueryRule(new ArrayList<QueryRule>());
+			disMaxQuery.setOperator(Operator.DIS_MAX);
+
+			List<String> synonyms = Lists.newArrayList(ontologyTerm.getSynonyms());
+			synonyms.add(ontologyTerm.getLabel());
+			for (String synonym : synonyms)
+			{
+				disMaxQuery.getNestedRules().add(
+						new QueryRule(AttributeMetaDataMetaData.LABEL, Operator.FUZZY_MATCH, synonym));
+			}
+
+			rules.add(disMaxQuery);
+		}
+
+		QueryRule finalDisMaxQuery = new QueryRule(rules);
+		finalDisMaxQuery.setOperator(Operator.DIS_MAX);
+
+		return finalDisMaxQuery;
+	}
+
+	public List<String> getAttributeIdentifiers(EntityMetaData sourceEntityMetaData)
+	{
+		Entity entityMetaDataEntity = dataService.findOne(EntityMetaDataMetaData.ENTITY_NAME,
+				new QueryImpl().eq(EntityMetaDataMetaData.FULL_NAME, sourceEntityMetaData.getName()));
+
+		if (entityMetaDataEntity == null) throw new MolgenisDataAccessException(
+				"Could not find EntityMetaDataEntity by the name of " + sourceEntityMetaData.getName());
+
+		return FluentIterable
+				.from(
+				entityMetaDataEntity.getEntities(EntityMetaDataMetaData.ATTRIBUTES)).transform(
+				new Function<Entity, String>()
+				{
+					public String apply(Entity attributeEntity)
+					{
+						return attributeEntity.getString(AttributeMetaDataMetaData.IDENTIFIER);
+					}
+				}).toList();
+	}
+
+	public List<OntologyTerm> findTagsSync(String description, List<String> ontologyIds)
```

Comment:
I'd name the Async method Async, if it exists, but otherwise all methods are Sync by default, so I'd not put that in the name.